File: backend/models/studio_models.py
"""
Studio Models - Notebook, Flashcards, Quiz
These models represent the core Studio features for NotebookPRO
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


# ============================================================================
# NOTEBOOK MODELS
# ============================================================================

class NotebookEntry(BaseModel):
    """A single note entry in the notebook"""
    id: str = Field(..., description="Unique identifier for the note")
    space_id: str = Field(..., description="Space this note belongs to")
    title: str = Field(..., description="Title of the note")
    content: str = Field(..., description="Main content/body of the note")
    source_type: str = Field(default="manual", description="Source: manual, chat, generated")
    source_id: Optional[str] = Field(None, description="ID of source (e.g., chat message ID)")
    tags: List[str] = Field(default_factory=list, description="Tags for categorization")
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class NotebookEntryCreate(BaseModel):
    """Request model for creating a notebook entry"""
    space_id: str
    title: str
    content: str
    source_type: str = "manual"
    source_id: Optional[str] = None
    tags: List[str] = []
    metadata: Dict[str, Any] = {}


class NotebookEntryUpdate(BaseModel):
    """Request model for updating a notebook entry"""
    title: Optional[str] = None
    content: Optional[str] = None
    tags: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


# ============================================================================
# FLASHCARD MODELS
# ============================================================================

class DifficultyLevel(str, Enum):
    """Difficulty level for flashcards"""
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class MasteryLevel(str, Enum):
    """User's mastery level for a flashcard"""
    NEW = "new"
    LEARNING = "learning"
    REVIEWING = "reviewing"
    MASTERED = "mastered"


class Flashcard(BaseModel):
    """A single flashcard for memorization"""
    id: str = Field(..., description="Unique identifier")
    space_id: str = Field(..., description="Space this flashcard belongs to")
    question: str = Field(..., description="Front of the card (question/prompt)")
    answer: str = Field(..., description="Back of the card (answer/explanation)")
    difficulty: DifficultyLevel = Field(default=DifficultyLevel.MEDIUM)
    topic: Optional[str] = Field(None, description="Topic matching the concept covered")
    mastery: MasteryLevel = Field(default=MasteryLevel.NEW)
    source_type: str = Field(default="manual", description="Source: manual, generated, notebook")
    source_id: Optional[str] = Field(None, description="Source ID (e.g., notebook entry ID)")
    tags: List[str] = Field(default_factory=list)
    review_count: int = Field(default=0, description="Number of times reviewed")
    correct_count: int = Field(default=0, description="Number of times answered correctly")
    
    # SM-2 Spaced Repetition Fields
    interval: int = Field(default=0, description="Interval in days before next review")
    repetitions: int = Field(default=0, description="Number of successful consecutive reviews")
    ease_factor: float = Field(default=2.5, description="SM-2 Ease factor")
    
    last_reviewed: Optional[datetime] = None
    next_review: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.now)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class FlashcardCreate(BaseModel):
    """Request model for creating a flashcard"""
    space_id: str
    question: str
    answer: str
    difficulty: DifficultyLevel = DifficultyLevel.MEDIUM
    source_type: str = "manual"
    source_id: Optional[str] = None
    tags: List[str] = []
    metadata: Dict[str, Any] = {}


class FlashcardUpdate(BaseModel):
    """Request model for updating a flashcard"""
    question: Optional[str] = None
    answer: Optional[str] = None
    difficulty: Optional[DifficultyLevel] = None
    mastery: Optional[MasteryLevel] = None
    tags: Optional[List[str]] = None


class FlashcardReview(BaseModel):
    """Request model for reviewing a flashcard"""
    quality: int = Field(..., description="0-5 score user gives after seeing answer (0-2 = forgot, 3 = hard, 4 = good, 5 = easy)")


class FlashcardGenerateRequest(BaseModel):
    """Request to generate flashcards from content"""
    space_id: str
    source_type: str = Field(..., description="Source type: notebook, file, text")
    source_ids: Optional[List[str]] = Field(None, description="IDs of notebook entries or files")
    text_content: Optional[str] = Field(None, description="Direct text content to generate from")
    num_cards: int = Field(default=5, description="Number of flashcards to generate")
    difficulty: DifficultyLevel = DifficultyLevel.MEDIUM


# ============================================================================
# QUIZ MODELS
# ============================================================================

class QuestionType(str, Enum):
    """Type of quiz question"""
    MULTIPLE_CHOICE = "multiple_choice"
    TRUE_FALSE = "true_false"
    SHORT_ANSWER = "short_answer"


class QuizQuestion(BaseModel):
    """A single quiz question"""
    id: str = Field(..., description="Unique identifier")
    question: str = Field(..., description="Question text")
    type: QuestionType = Field(..., description="Question type")
    options: Optional[List[str]] = Field(None, description="Options for multiple choice")
    correct_answer: str = Field(..., description="Correct answer")
    explanation: Optional[str] = Field(None, description="Explanation of the answer")
    points: int = Field(default=1, description="Points for this question")
    difficulty: DifficultyLevel = Field(default=DifficultyLevel.MEDIUM)


class Quiz(BaseModel):
    """A quiz session"""
    id: str = Field(..., description="Unique identifier")
    space_id: str = Field(..., description="Space this quiz belongs to")
    title: str = Field(..., description="Quiz title")
    description: Optional[str] = None
    questions: List[QuizQuestion] = Field(..., description="List of questions")
    source_type: str = Field(default="manual", description="Source: manual, generated, notebook, file")
    source_ids: Optional[List[str]] = None
    created_at: datetime = Field(default_factory=datetime.now)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class QuizCreate(BaseModel):
    """Request model for creating a quiz"""
    space_id: str
    title: str
    description: Optional[str] = None
    questions: List[QuizQuestion] = []
    source_type: str = "manual"
    source_ids: Optional[List[str]] = None


class QuizGenerateRequest(BaseModel):
    """Request to generate a quiz from content"""
    space_id: str
    title: str
    source_type: str = Field(..., description="Source type: notebook, file, text")
    source_ids: Optional[List[str]] = Field(None, description="IDs of notebook entries or files")
    text_content: Optional[str] = Field(None, description="Direct text content")
    num_questions: int = Field(default=5, description="Number of questions")
    question_types: List[QuestionType] = Field(
        default=[QuestionType.MULTIPLE_CHOICE],
        description="Types of questions to include"
    )
    difficulty: DifficultyLevel = DifficultyLevel.MEDIUM


class QuizAnswer(BaseModel):
    """User's answer to a quiz question"""
    question_id: str
    answer: str
    time_spent: Optional[int] = Field(None, description="Time spent in seconds")


class QuizSubmission(BaseModel):
    """User's quiz submission"""
    quiz_id: str
    answers: List[QuizAnswer]


class QuizResult(BaseModel):
    """Result of a quiz submission"""
    quiz_id: str
    submission_id: str = Field(..., description="Unique submission ID")
    total_questions: int
    correct_answers: int
    incorrect_answers: int
    score_percentage: float
    total_points: int
    earned_points: int
    answers: List[Dict[str, Any]] = Field(..., description="Detailed answer results")
    completed_at: datetime = Field(default_factory=datetime.now)
    time_taken: Optional[int] = Field(None, description="Total time in seconds")


class QuizHistory(BaseModel):
    """Quiz attempt history"""
    quiz_id: str
    space_id: str
    quiz_title: str
    results: List[QuizResult] = Field(default_factory=list)
    best_score: float = Field(default=0.0)
    average_score: float = Field(default=0.0)
    attempts_count: int = Field(default=0)

File: notebook_pro_app/lib/models/studio_models.dart
// Studio Models - Notebook, Flashcards, Quiz
// These models match the backend Pydantic models

// ============================================================================
// NOTEBOOK MODELS
// ============================================================================

class NotebookEntry {
  final String id;
  final String spaceId;
  final String title;
  final String content;
  final String sourceType;
  final String? sourceId;
  final List<String> tags;
  final DateTime createdAt;
  final DateTime updatedAt;
  final Map<String, dynamic> metadata;

  NotebookEntry({
    required this.id,
    required this.spaceId,
    required this.title,
    required this.content,
    this.sourceType = 'manual',
    this.sourceId,
    this.tags = const [],
    required this.createdAt,
    required this.updatedAt,
    this.metadata = const {},
  });

  factory NotebookEntry.fromJson(Map<String, dynamic> json) {
    return NotebookEntry(
      id: json['id'],
      spaceId: json['space_id'],
      title: json['title'],
      content: json['content'],
      sourceType: json['source_type'] ?? 'manual',
      sourceId: json['source_id'],
      tags: List<String>.from(json['tags'] ?? []),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      metadata: json['metadata'] ?? {},
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'space_id': spaceId,
      'title': title,
      'content': content,
      'source_type': sourceType,
      'source_id': sourceId,
      'tags': tags,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'metadata': metadata,
    };
  }
}

// ============================================================================
// FLASHCARD MODELS
// ============================================================================

enum DifficultyLevel {
  easy,
  medium,
  hard;

  String toJson() => name;

  static DifficultyLevel fromJson(String value) {
    return DifficultyLevel.values.firstWhere((e) => e.name == value);
  }
}

enum MasteryLevel {
  newCard('new'),
  learning('learning'),
  reviewing('reviewing'),
  mastered('mastered');

  final String value;
  const MasteryLevel(this.value);

  String toJson() => value;

  static MasteryLevel fromJson(String value) {
    return MasteryLevel.values.firstWhere((e) => e.value == value);
  }
}

class Flashcard {
  final String id;
  final String spaceId;
  final String question;
  final String answer;
  final DifficultyLevel difficulty;
  final MasteryLevel mastery;
  final String sourceType;
  final String? sourceId;
  final List<String> tags;
  final int reviewCount;
  final int correctCount;
  final DateTime? lastReviewed;
  final DateTime? nextReview;
  final DateTime createdAt;
  final Map<String, dynamic> metadata;

  Flashcard({
    required this.id,
    required this.spaceId,
    required this.question,
    required this.answer,
    this.difficulty = DifficultyLevel.medium,
    this.mastery = MasteryLevel.newCard,
    this.sourceType = 'manual',
    this.sourceId,
    this.tags = const [],
    this.reviewCount = 0,
    this.correctCount = 0,
    this.lastReviewed,
    this.nextReview,
    required this.createdAt,
    this.metadata = const {},
  });

  factory Flashcard.fromJson(Map<String, dynamic> json) {
    return Flashcard(
      id: json['id'],
      spaceId: json['space_id'],
      question: json['question'],
      answer: json['answer'],
      difficulty: DifficultyLevel.fromJson(json['difficulty'] ?? 'medium'),
      mastery: MasteryLevel.fromJson(json['mastery'] ?? 'new'),
      sourceType: json['source_type'] ?? 'manual',
      sourceId: json['source_id'],
      tags: List<String>.from(json['tags'] ?? []),
      reviewCount: json['review_count'] ?? 0,
      correctCount: json['correct_count'] ?? 0,
      lastReviewed: json['last_reviewed'] != null
          ? DateTime.parse(json['last_reviewed'])
          : null,
      nextReview: json['next_review'] != null
          ? DateTime.parse(json['next_review'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
      metadata: json['metadata'] ?? {},
    );
  }

  double get accuracy =>
      reviewCount > 0 ? (correctCount / reviewCount * 100) : 0.0;
}

// ============================================================================
// QUIZ MODELS
// ============================================================================

enum QuestionType {
  multipleChoice('multiple_choice'),
  trueFalse('true_false'),
  shortAnswer('short_answer');

  final String value;
  const QuestionType(this.value);

  String toJson() => value;

  static QuestionType fromJson(String value) {
    return QuestionType.values.firstWhere((e) => e.value == value);
  }
}

class QuizQuestion {
  final String id;
  final String question;
  final QuestionType type;
  final List<String>? options;
  final String correctAnswer;
  final String? explanation;
  final int points;
  final DifficultyLevel difficulty;

  QuizQuestion({
    required this.id,
    required this.question,
    required this.type,
    this.options,
    required this.correctAnswer,
    this.explanation,
    this.points = 1,
    this.difficulty = DifficultyLevel.medium,
  });

  factory QuizQuestion.fromJson(Map<String, dynamic> json) {
    return QuizQuestion(
      id: json['id'],
      question: json['question'],
      type: QuestionType.fromJson(json['type'] ?? 'multiple_choice'),
      options: json['options'] != null
          ? List<String>.from(json['options'])
          : null,
      correctAnswer: json['correct_answer'],
      explanation: json['explanation'],
      points: json['points'] ?? 1,
      difficulty: DifficultyLevel.fromJson(json['difficulty'] ?? 'medium'),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'question': question,
      'type': type.toJson(),
      'options': options,
      'correct_answer': correctAnswer,
      'explanation': explanation,
      'points': points,
      'difficulty': difficulty.toJson(),
    };
  }
}

class Quiz {
  final String id;
  final String spaceId;
  final String title;
  final String? description;
  final List<QuizQuestion> questions;
  final String sourceType;
  final List<String>? sourceIds;
  final DateTime createdAt;
  final Map<String, dynamic> metadata;

  Quiz({
    required this.id,
    required this.spaceId,
    required this.title,
    this.description,
    required this.questions,
    this.sourceType = 'manual',
    this.sourceIds,
    required this.createdAt,
    this.metadata = const {},
  });

  factory Quiz.fromJson(Map<String, dynamic> json) {
    return Quiz(
      id: json['id'],
      spaceId: json['space_id'],
      title: json['title'],
      description: json['description'],
      questions: (json['questions'] as List)
          .map((q) => QuizQuestion.fromJson(q))
          .toList(),
      sourceType: json['source_type'] ?? 'manual',
      sourceIds: json['source_ids'] != null
          ? List<String>.from(json['source_ids'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
      metadata: json['metadata'] ?? {},
    );
  }

  int get totalPoints => questions.fold(0, (sum, q) => sum + q.points);
}

class QuizAnswer {
  final String questionId;
  final String answer;
  final int? timeSpent;

  QuizAnswer({
    required this.questionId,
    required this.answer,
    this.timeSpent,
  });

  Map<String, dynamic> toJson() {
    return {
      'question_id': questionId,
      'answer': answer,
      'time_spent': timeSpent,
    };
  }
}

class QuizResult {
  final String quizId;
  final String submissionId;
  final int totalQuestions;
  final int correctAnswers;
  final int incorrectAnswers;
  final double scorePercentage;
  final int totalPoints;
  final int earnedPoints;
  final List<Map<String, dynamic>> answers;
  final DateTime completedAt;
  final int? timeTaken;

  QuizResult({
    required this.quizId,
    required this.submissionId,
    required this.totalQuestions,
    required this.correctAnswers,
    required this.incorrectAnswers,
    required this.scorePercentage,
    required this.totalPoints,
    required this.earnedPoints,
    required this.answers,
    required this.completedAt,
    this.timeTaken,
  });

  factory QuizResult.fromJson(Map<String, dynamic> json) {
    return QuizResult(
      quizId: json['quiz_id'],
      submissionId: json['submission_id'],
      totalQuestions: json['total_questions'],
      correctAnswers: json['correct_answers'],
      incorrectAnswers: json['incorrect_answers'],
      scorePercentage: (json['score_percentage'] as num).toDouble(),
      totalPoints: json['total_points'],
      earnedPoints: json['earned_points'],
      answers: List<Map<String, dynamic>>.from(json['answers']),
      completedAt: DateTime.parse(json['completed_at']),
      timeTaken: json['time_taken'],
    );
  }

  bool get passed => scorePercentage >= 60.0;
}
